azurelinuxagent/common/protocol/extensions_goal_state.py

```diff
@@ -97,29 +98,68 @@ def compare(from_extensions_config, from_vm_settings):
 
         NOTE: The order of the two instances is important for the debug info to be logged correctly (ExtensionsConfig first, vmSettings second)
         """
-        def compare_attribute(attribute):
-            from_extensions_config_value = getattr(from_extensions_config, attribute)
-            from_vm_settings_value = getattr(from_vm_settings, attribute)
-
-            if from_extensions_config_value != from_vm_settings_value:
-                message = "Mismatch in ExtensionsConfig (incarnation {0}) and vmSettings (etag {1}).\nAttribute: {2}()\n{3} != {4}".format(
-                    from_extensions_config.id, from_vm_settings.id,
-                    attribute,
-                    from_extensions_config_value, from_vm_settings_value
-                )
-                raise GoalStateMismatchError(message)
-
-        compare_attribute("activity_id")
-        compare_attribute("correlation_id")
-        compare_attribute("created_on_timestamp")
-        # The status blob was added after version 112
-        if from_vm_settings.host_ga_plugin_version > FlexibleVersion("1.0.8.112"):
-            compare_attribute("status_upload_blob")
-            compare_attribute("status_upload_blob_type")
-        compare_attribute("required_features")
-        compare_attribute("on_hold")
-        compare_attribute("agent_manifests")
-        compare_attribute("extensions")
+        context = []  # used to keep track of the attribute that is being compared
```

Comment:
Initially I implemented this method to try to catch errors in the HostGAPlugin and it was pretty simple, since I do not expect the HostGAPlugin to return incorrect values deep in the goal state's object model.

However, while parsing the extension settings from vmSettings, I realized this method is extremely useful to compare against the current parsing of extensionsConfig,

This new implementation compares every single attribute of the goal state's object model and adds extra context information to debug differences in the parsing.

(while writing this code I found some subtle details wrt representing numeric attributes a strings or ints, empty collections or None, etc)